cep-oci.md

```diff
@@ -0,0 +1,157 @@
+
+The manifest for a conda package on an OCI registry should look like follows.
+
+It should have a name and a tag. The name is `<channel>/<subdir>/<package-name>`.
+The tag is the version and build string of the packages, using a `-` as a separator.
+
+For example, a package like `xtensor-0.10.4-h431234.conda` would map to a OCI registry `conda-forge/linux-64/xtensor:0.10.4-h431234`.
+
+A conda package, in an OCI registry, should ship up to 3 layers:
+
+- The package data itself, as a tarball. (mandatory)
+  - This can be either a `.tar.bz2` (v1) or a `.conda` (v2) file, or both as separate layers.
+- The package `info` folder as a gzipped "tar.gz" file.
+- The package `info/index.json` file as a plain JSON file.
+
+### Mapping a conda-package to the OCI registry
+
+A given conda-package is identified by a URL like `<subdir>/<package-name>-<version>-<build>.<ext>` where `<subdir>` is the platform and architecture, `<package-name>` is the name of the package, `<version>` is the version of the package, `<build>` is the build string of the package, and `<ext>` is the extension of the package file.
+
+#### Mapping the package name
+
+> [!NOTE]
+> **Package names in the conda world**
+> 
+> The following regex is given by [`conda/schemas`](https://github.com/conda/schemas/blob/473708ac97283708d6664cbd89b8049ad1623489/common-1.schema.json#L58-L82) for a valid package name: `^[a-z0-9_](?!_)[._-]?([a-z0-9]+(\.|-|_|$))*$`
+> 
+> That means, a package can start with an alphanumeric character or a _single_ underscore (not multiple), and can contain dots, dashes, and underscores. It also has to end with a alphanumeric character (cannot end with a dot, dash, or underscore).
+
+To store this package on an OCI registry, we need to map it to a name and tag. The name is `<channel>/<subdir>/<package-name>`. The tag is `<version>-<build>`. There are some special rules for OCI registry names and tags for which we need some mapping. The regex for valid names is as follows:
+
+`[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*(\/[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*)*` ([ref](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pulling-manifests))
+
+The regex expresses that names can only start with an alphanumeric letter.
+
+In `conda`, names can start with an underscore and it is used by conda-forge (e.g. `_libgcc_mutex`). For this reason, we prepend packages with a leading underscore with the string `zzz`. The name would thus be changed to `zzz_libgcc_mutex`.
+
+#### Mapping the tag
+
+The tag is the version and build string of the packages, using a `-` as a separator. However, a OCI tag has to conform to the following regex:
+
+`[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}`
+
+Some characters that are used in the conda-forge repository as part of the build string are not allowed in the OCI registry. For this reason, we use the following mapping:
+
+- `+` is replaced by `__p__`
+- `!` is replaced by `__e__`
+- `=` is replaced by `__eq__`
```

Comment:
That list seems to be a little random, it looks like Python dunder methods, but not quite valid, how about this instead?

```suggestion
- `+` is replaced by `__add__`
- `!` is replaced by `__not__`
- `=` is replaced by `__eq__`
```

Reply:
`p` for `plus` and `e` for `exclamation`, I guess? These characters might not mean "adding" or "negating" in a version/build string, so I'm not sure whether this is a good way to solve the "randomness" here. These names are only for the OCI names, anyway, right? They are renamed once downloaded.

Reply:
> `p` for `plus` and `e` for `exclamation`, I guess? 

Yes!
These mappings (prepending the packages names starting with `_` with `zzz` and replacing `+` etc for tags) are really only intended to make them valid in order to be stored in an OCI registry (correct me if I'm wrong @wolfv). It doesn't impact anything elsewhere as the names and tags are still visible (by users when downloaded for example) as their original strings.
So I don't think having `internal` instead of `zzz` (same for the tags suggestions) would really matter (apart from making sure that it wouldn't conflict with potential similar existing strings).
While something more explicit would be preferable,  given that a bunch of packages have already been mirrored that way as @jaimergp  mentioned, I would say to keep it as is?

Reply:
Is there a standard encoding scheme we can rely on here instead of inventing our own?

Reply:
Also, if we insist of a special encoding, we can add underscores themselves as `__under__`, which would render our scheme unambiguous, since an existing package with `__under__` in it would be mapped to five `__under__` in a row.

Reply:
@beckermr - underscores are allowed, but not at the beginning of the package name. Not sure what you mean. 

We could encode the strings with something like SHA256 or MD5 but then they are not easily recoverable and more cryptic. We selected this "encoding" to make it work, while still possible to map back (even as a human) to the source string. I am not particularly attached to this encoding, so very open to alternative suggestions.

- We could remove the offending characters (but then it might not be unique per package name anymore, so might not work
- We could append a hash, and remove the offending characters. That should work fine, but look weird, too

Reply:
> Is there a standard encoding scheme we can rely on here instead of inventing our own?

The text itself mentiones that package names can start with an underscore. _libgcc_mutex is the example given.

Perhaps we can learn from symbol name mangling instead of coming up with a unique scheme?

Reply:
Another interesting thing from the [spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pulling-manifests):

> Implementers note: Many clients impose a limit of 255 characters on the length of the concatenation of the registry hostname (and optional port), /, and <name> value. If the registry name is registry.example.org:5000, those clients would be limited to a <name> of 229 characters (255 minus 25 for the registry hostname and port and minus 1 for a / separator). For compatibility with those clients, registries should avoid values of <name> that would cause this limit to be exceeded.

Do we see this as a risk for hosting channels with long names? Does this character limit only apply for fetching the manifest or also artifacts in real world OCI implementations?

Reply:
@wolfv I made a mistake above, but the point stands. Suppose we have two packages:

 - zzzblah
 - _blah

The current scheme replacing only leading underscores with `zzz` would render the packages as 

 - zzzblah
 - zzzblah

creating a conflict.

However, if we encode underscores with a pattern that uses underscores and always encode all of them, say `under__`, then for a potential conflict, we always mangle the encoding string too. For example, with

 - under__blah
 - _blah

we get

 - underunder__under__blah
 - under__blah

The mistake I made above was using a pattern with a leading underscore for an underscore. That won't work.

Reply:
Then we need to forbid using `under_` as a part of package name in conda because if we have a package named `under__something` (notice the extra underscore), we won't know if the encoded result `under_under_` is actually `_under_something`, `under__something` or `__something` :D

Reply:
The other question: do we need to map back? Or can we just check the internal metadata at that point? I think we just need to go back from `original` to `mangled`, but not back, right? Once we have fetched `mangled` we can check if the metadata matches `original`. So I think we are fine as long as original maps to mangled in a unique way, and that might mean "exploding" all the original underscores into `under_`.

Reply:
> Then we need to forbid using under_ as a part of package name in conda because if we have a package named under__something (notice the extra underscore), we won't know if the encoded result under_under_ is actually _under_something, under__something or __something :D

No. @jaimergp  `under__something` gets translated to `underunder__under__somthing`. **Every underscore gets translated.**

Reply:
I have not. The issue with any scheme is that it cannot start with an `_` because the OCI package name cannot start with an `_`.

Reply:
Reading all of the above, I think we want:

1. Each package should have a unique representation when converted to a format that is accepted in OCI (I will refer to it as mangled).
2. Mangled package names should be kept as short as possible because there is a limit on the number of characters (128).
3. The mangled names should not stray too far away from the unmangled human-readable text to make it easy on users when navigating an OCI registry (otherwise base64 encoding would also suffice). 

So how about the following:

* We prepend all package names with a `c`. Packages that start with an underscore instead start with an uppercase `C`.     
* We escape any character that is not allowed with an underscore including the underscore itself.

    `+` -> `_P`
    `!` -> `_N`
    `=` -> `_E`
    `_` -> `__`
    
    I used uppercase letters here because they standout between the usually lowercase text.
    
E.g: 

    `conda` -> `cconda`
    `_libgcc_mutex` -> `Clibgcc_mutex`
     
Tags:

* `1!0.0.0_dev0+placeholder` -> `1_N0.0.0__dev0_Pplaceholder`

Thoughts?

Reply:
@beckermr - when writing out on the file system we would use the real name anyways (not the mapped name).

Although I would have to double-check if uppercase is allowed in OCI registries, and wether the registry is case-sensitive.

Reply:
>This will not work on case-insensitive file systems, which includes OSX in some cases.

I added case-insensitivity as a requirement and modified the example.

Reply:
@baszalmstra It appears that lower-case only is required by the OCI spec for container names. The names must match this regex `[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*(\/[a-z0-9]+((\.|_|__|-+)[a-z0-9]+)*)*` [here](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pull) IIUIC.

Reply:
Tags can have upper case though. The regex is `[a-zA-Z0-9_][a-zA-Z0-9._-]{0,127}` from the same document.

Reply:
Are conda package names case sensitive?

Reply:
Based on your findings do you want to change anything with regard to the name mangling suggested above?

Reply:
Conda package names are always lower case, so presumably case insensitive.

Reply:
@baszalmstra we might as well have everything lowercase since that is what the OCI registry will look like.

Reply:
Case insensitive filesystems are usually case preserving though.